extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/conf/LdapBindGuacamoleProperty.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.auth.ldap.conf;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.directory.api.ldap.model.name.Dn;
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.GuacamoleServerException;
+import org.apache.guacamole.properties.GuacamoleProperty;
+
+/**
+ * A GuacamoleProperty that checks a string to a verify that it is a valid
+ * LDAP bind format - either an Active Directory UPN or a LDAP DN.  An exception
+ * is thrown if the provided value is invalid.
+ */
+public abstract class LdapBindGuacamoleProperty implements GuacamoleProperty<String> {
+
+    /**
+     * An expression that matches valid Active Directory UPN values, in one of
+     * three formats - <DOMAIN>\<USERNAME>, <USERNAME>@<DOMAIN> or <USERNAME>
+     */
+    private static final Pattern LDAP_UPN_PATTERN = Pattern.compile("^(?:(?<adusername>)([^@]+(?<addomain>.*))|(?<nbdomain>[^\\\\]+)\\\\(?<nbusername>.+))$");
```

Comment:
I've been playing around with this regex and various test strings, and the regex itself is actually pretty broad. The documentation for the regex sets an honorable goal, but the only strings that will not match this are strings that start with "@" but do not end with "\something".

Checking [Microsoft's documentation for the UPN](https://docs.microsoft.com/en-us/windows/win32/adschema/a-userprincipalname), it's clear that there is an intent for the value to follow a specific format ("based on" RFC 822 and "should" be the user's email address), but there is no apparent hard requirement except that it must be a Unicode string.

Also, while Active Directory may specifically allow username-style binds driven by `userPrincipalName`, I'm not sure this goes for other LDAP server implementations that may have their own definitions for what usernames can be used in lieu of an FQDN.

I think we should probably stick to a plain `StringGuacamoleProperty`...

Reply:
Yeah, that makes sense. I've reverted it to just a `StringGuacamoleProperty`.